substrate/frame/staking/stake-tracker/src/lib.rs

```diff
@@ -0,0 +1,543 @@
+							let _ = T::TargetList::on_update(who, balance).defensive_proof(
+								"staker exists in the list as per the check above; qed.",
+							);
+						}
+					} else {
+						defensive!("unexpected: unable to fetch score from staking interface of an existent staker");
+					}
+				},
+			};
+		}
+
+		// ------ Helpers
+
+		/// Helper to convert the balance of a staker into its vote weight.
+		pub(crate) fn to_vote(balance: BalanceOf<T>) -> VoteWeight {
+			<T::Staking as StakingInterface>::CurrencyToVote::to_vote(
+				balance,
+				T::Currency::total_issuance(),
+			)
+		}
+
+		/// Helper to fetch te active stake of a staker and convert it to vote weight.
+		pub fn vote_of(who: &T::AccountId) -> VoteWeight {
+			let active = T::Staking::stake(who).map(|s| s.active).defensive_unwrap_or_default();
+			Self::to_vote(active)
+		}
+
+		/// Returns whether a nomination vec has duplicate targets.
+		///
+		/// Used for debug assertions only, since this pallet expects the nominations to be
+		/// deduplicated at all places.
+		pub fn has_duplicate_nominations(mut v: Vec<T::AccountId>) -> bool {
+			use sp_std::collections::btree_set::BTreeSet;
+			let size_before = v.len();
+			let dedup = v.drain(..).collect::<BTreeSet<_>>().into_iter().collect::<Vec<_>>();
+
+			size_before != dedup.len()
+		}
+	}
+}
+
+impl<T: Config> OnStakingUpdate<T::AccountId, BalanceOf<T>> for Pallet<T> {
+	/// When a nominator's stake is updated, all the nominated targets must be updated
+	/// accordingly.
+	///
+	/// The score of the node associated with `who` in the *VoterList* will be updated if the
+	/// the mode is [`VoterUpdateMode::Strict`]. The approvals of the nominated targets (by `who`)
+	/// are always updated.
+	fn on_stake_update(
+		who: &T::AccountId,
+		prev_stake: Option<Stake<BalanceOf<T>>>,
+		stake: Stake<BalanceOf<T>>,
+	) {
+		match T::Staking::status(who) {
+			Ok(StakerStatus::Nominator(nominations)) =>
+				Self::do_stake_update_voter(who, prev_stake, stake, nominations),
+			Ok(StakerStatus::Validator) => Self::do_stake_update_target(who, prev_stake, stake),
+			Ok(StakerStatus::Idle) => (), // nothing to see here.
+			Err(_) => {
+				defensive!(
+					"staker should exist when calling `on_stake_update` and have a valid status"
+				);
+			},
+		}
+	}
+
+	/// Triggered when a new validator is added to the system.
+	///
+	/// Overview: If `who` is part of the target list, update its score. Otherwise, insert a new
+	/// node to the target list with self-stake as initial node score.
+	///
+	/// A validator is also considered a voter with self-vote and should be added to
+	/// [`Config::VoterList`].
+	fn on_validator_add(who: &T::AccountId, self_stake: Stake<BalanceOf<T>>) {
+		let self_stake = Self::to_vote(self_stake.active).into();
+
+		match T::TargetList::on_insert(who.clone(), self_stake) {
+			Ok(_) => (),
+			Err(_) => {
+				// if the target already exists in the list, it means that the target is idle
```

Comment:
Correct, a dangling target that reset the `validate` intention. In this case, we add the self-stake to the already existing score in the target list (because the target was dangling). I made the comments more explicit.